controllers/todo.js

```diff
@@ -1,40 +1,173 @@
+  return false;
+}
+
 const getParticularToDo = async (req, res) => {
-  // Get the Todo of the logged in user with given id.
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+  // console.log(creator(userId, todo), collaborator(userId, todo));
+
+  if (!todo) {
+    res.status(404).send("Todo of given id not found!")
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    res.status(200).json(todo);
+  } else {
+    res.status(403).send("Logged in user not authorized to view todo of given id.");
+  }
 };
 
 const editToDo = async (req, res) => {
-  // Change the title of the Todo with given id, and get the new title as response.
+  const { title } = req.body;
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+
+  if (!title) {
+    res.status(400).send("Title field is empty!");
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    const queryTodo = await ToDo.findByIdAndUpdate(todoId, { title: title });
+
+    if (queryTodo) {
+      res.status(200).json({ id: todoId, title: title });
+    } else {
+      res.status(404).send("Todo of given id not found!");
+    }
+  } else {
+    res.status(403).send("Logged user is not authorized to edit todo of given id!");
+  }
 };
 
 const editToDoPatch = async (req, res) => {
-  // Change the title of the Todo with given id, and get the new title as response
+  const { title } = req.body;
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+
+  if (!title) {
+    res.status(400).send("Title field is empty!");
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    const queryTodo = await ToDo.findByIdAndUpdate(todoId, { title: title });
+
+    if (queryTodo) {
+      res.status(200).json({ id: todoId, title: title });
+    } else {
+      res.status(404).send("Todo of given id not found!");
+    }
+  } else {
+    res.status(403).send("Logged user is not authorized to edit todo of given id!");
+  }
 };
 
 const deleteToDo = async (req, res) => {
-  //  Delete the todo with given id
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+
+  if (!todo) {
+    res.status(404).send("Todo of given id not found!");
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    const deleteTodo = await ToDo.findByIdAndDelete(todoId);
+
+    if (deleteTodo) {
+      res.status(204).send("Todo of given id deleted");
+    } else {
+      res.status(501).send("Couldn't delete task.");
+    }
+  } else {
+    res.status(403).send("Logged user is not authorized to edit todo of given id!");
+  }
 };
 
+const addCollaborator = async (req, res) => {
+  const { collaboratorId } = req.body;
+  const todoId = req.params.id;
+  const userId = req.user.id;
+
+  const collaborator = await User.findById(collaboratorId);
```

Comment:
Try giving a username and then finding the username by that id. That makes the addition and deletion much easier for a tester.